### ISSUE TITLE
StreamPeer::get_<NumberFormat>() method ignore error from get_data()

### ISSUE THREAD

**Issue by AnmAtAnm** [godotengine/godot#45518]
**Godot version:**
Both `3.2` and `master` branches

**Issue description:**
In the following WAT test, `get_8()` returns values beyond the end of the buffer:

```gdscript
func test_end_of_buffer():
    var stream = StreamPeerBuffer.new()
    stream.data_array = PoolByteArray([1,2,3])

    asserts.is_equal(1, stream.get_u8(), "First u8 value")
    asserts.is_equal(2, stream.get_u8(), "Second u8 value")
    asserts.is_equal(3, stream.get_u8(), "Third u8 value")
    # End of buffer

    # What? Why? 0 is a valid value!
    asserts.is_equal(0, stream.get_u8(), "EOF u8 value") 
```

Looking at the source of `core/io/stream_peer.cpp` ([master](https://github.com/godotengine/godot/blob/master/core/io/stream_peer.cpp#L225) and [3.2 branch](https://github.com/godotengine/godot/blob/3.2/core/io/stream_peer.cpp#L235):

```cpp
uint8_t StreamPeer::get_u8() {
	uint8_t buf[1];
	get_data(buf, 1);
	return buf[0];
}
```

`get_data(..)` returns an error code, but it is thrown away, and `buf` is never written. The code is near identical for all `get_number(..)` methods.

As written, there is no way to determine if a EOF is encountered on any StreamPeer implementation in one of these methods.

**Possible Solution:**
 At the very least, the error code should be stored and accessible.  Perhaps a read_only `last_error` property.

**Comment by AnmAtAnm** [comment]
Ideally all `get_*()` methods would behave identically.  Currently, methods like `get_String()` will `ERR_FAIL_COND_V(err != OK, ...)`.  This is a forced crash, right? 

If so, that isn't fair to the developer to know where the IO stream will end, particularly in with respect to how the string will be deserialized. Consider `get_utf8_string()` where the length in bytes varies, even for the same number of code points.

So I think this is another case where something like a `last_error` property is a better solution.

**Comment by akien-mga** [comment]
> Currently, methods like `get_String()` will `ERR_FAIL_COND_V(err != OK, ...)`. This is a forced crash, right?

That's not a crash, it's just an error print and return with the specified value. It doesn't let you handle it programmatically though.

**Comment by Faless** [comment]
Not fond of `last_error`. Would be pretty inconsistent with the rest of the Godot API.
We could add `ERR_FAIL`s to at print the error though.

> If so, that isn't fair to the developer to know where the IO stream will end

`stream.get_size() - stream.get_position()` gives you the current position in the stream. If the result is 0, you reached the end of the stream

**Comment by AnmAtAnm** [comment]
> `stream.get_size() - stream.get_position()` gives you the current position in the stream. If the result is 0, you reached the end of the stream

Two issues:
 1. `get_size()` and `get_position()` are specific to `StreamPeerBuffer`, but the underlying problem is with `StreamPeer` implementation.
 2. There is no distinction between a successful read that happens to align with EOF, and a read of multiple bytes that extends beyond the EOF.

**Comment by Faless** [comment]
@AnmAtAnm which `StreamPeer` are you using?

**Comment by AnmAtAnm** [comment]
@Faless At the moment, I'm using `StreamPeerBuffer`, but I expect to process networked data with my classes at some point. Why is it relevant?

**Comment by AnmAtAnm** [comment]
@Faless 
> Not fond of `last_error`. Would be pretty inconsistent with the rest of the Godot API.

Agreed that it is different, but I don't have a better idea for reading primitives in GDScript. The reference parameter used by the underlying `get_data(...)` is not supported by the language. Our only out parameter for primitives is either the return value or a capturing array passed in (which seems like an uglier solution).

The major benefits of the `last_error` are
 * It is does not break the existing API (`get_number()` methods can continue to return the exact same values), thus can be patched on top of the 3.2 branch.
 * The same pattern can be applied to all `get_*()` methods, even if they already return an error, thus unifying the error checking for the `StreamPeer` methods.

**Comment by AnmAtAnm** [comment]
Let me just recognize here that one can avoid getting into the error state by checking `get_available_bytes()`:

```gdscript
if stream.get_available_bytes() < 8:
    handle_eof()
    # and maybe return
var value = stream.get_64()
```

This does mean the developer has to know how many bytes are expected for a cluster of reads. It only covers the EOF case, where last_error provides a clear distinguish errors. For example:
* `get_string(-1)` might `ERR_OUT_OF_MEMORY` before reaching the end of a file
* `get_string_utf8(n)` might have an incomplete code point just before the EOF or allotted bytes (possibly returning the string read thus far)

Additionally, storing `last_error` with the stream can simplify the code around error checking, by centralizing it further up the call stack:

```gdscript
var data = read_struct_of_unknown_size(stream)
if stream.last_error:
    handle_error()
```
